### PR TITLE
OCPBUGS-12885: daemon: stop using `nmstatectl persist-nic-names --inspect` on el9

### PR DESCRIPTION
**Pull request by jlebon**
Support for that switch was dropped.[[1]]

[1]: https://github.com/nmstate/nmstate/pull/2306

(And other minor prep patches split out of https://github.com/openshift/machine-config-operator/pull/3684.)

### CODE REVIEW

**Review by yuqi-zhang — APPROVED** [review]
Logically seems fine, this should only run when a rhel8 based bootimage tries to join the cluster right?

Leaving a

/hold

for pre-merge verification.

**Comment by sinnykumari** [comment]
Thanks @jlebon for the putting up the fix. Should we attach this PR to the bug so that this can also be backported to 4.13?

**Review by sinnykumari — COMMENTED** [review]
https://github.com/openshift/machine-config-operator/blob/master/templates/common/_base/units/machine-config-daemon-firstboot.service.yaml#L19 also needs to be updated to use `--persist-nics` flag instead of `--maybe-persist-nics`

**Comment by jlebon** [comment]
> [`master`/templates/common/_base/units/machine-config-daemon-firstboot.service.yaml#L19](https://github.com/openshift/machine-config-operator/blob/master/templates/common/_base/units/machine-config-daemon-firstboot.service.yaml?rgh-link-date=2023-04-24T15%3A03%3A25Z#L19) also needs to be updated to use `--persist-nics` flag instead of `--maybe-persist-nics`

Thanks, updated!

**Comment by sinnykumari** [comment]
Backport to 4.13 when this PR gets merged.
/cherry-pick release-4.13

**Comment by jlebon** [comment]
> Logically seems fine, this should only run when a rhel8 based bootimage tries to join the cluster right?

It will also run in the upgrade path from 4.12 to 4.13 when the 4.13 MCO starts up before it updates the el8 host to el9.

**Comment by jlebon** [comment]
Sorry, I had to push another update here to fix a subtle issue I noticed in the last commit.

**Comment by jlebon** [comment]
> > Backport to 4.13 when this PR gets merged.
> > /cherry-pick release-4.13
> 
> Note that backporting this as is would risk breaking 4.13 transitively too. We could have the PR ready, and then build nmstate and once it's built, hit merge on the PR.

Note 4.13 is already broken; the nmstate patch that removes `--inspect` is already in the latest 4.13 RHCOS builds.

> But I would have made the PR handle both old and new nmstate interfaces.

Yeah agreed, that's probably what we should've done. Since 4.13 is already broken anyway, seems like we might as well just drop support for `--inspect` at this point?

**Comment by sdodson** [comment]
> Note 4.13 is already broken; the nmstate patch that removes `--inspect` is already in the latest 4.13 RHCOS builds. (Or does "transitively" here refer to some other kind of breakage?)


I don't actually see nmstate in RHCOS. Is that a pending change? It seems to be included in MCD image.

**Comment by jlebon** [comment]
> > Note 4.13 is already broken; the nmstate patch that removes `--inspect` is already in the latest 4.13 RHCOS builds. (Or does "transitively" here refer to some other kind of breakage?)
> 
> I don't actually see nmstate in RHCOS. Is that a pending change? It seems to be included in MCD image.
> 
> edit: looks like `nmstate-2.2.7` is in 4.14-9.2 builds but not 4.13 yet. MCD image has `nmstate-2.2.9-4.rhaos4.13.el8` which I assume is intentionally an el8 build to ensure compatibility running on RHEL8 hosts?

Right thanks, I got the details wrong:
- RHCOS 4.14 ships nmstate (inherited from FCOS it seems [which added it recently](https://github.com/coreos/fedora-coreos-config/pull/2269))
- RHCOS 4.13 doesn't ship nmstate. I don't *think* there are plans to change that.
- None of that matters for this; we ship nmstate in the MCD image directly.
- The breakage therefore would be due to nmstate-2.2.9-4.rhaos4.13.el8 entering the 4.13 MCD image.

If there's an easy lever for this, the fastest fix would probably be to untag that build and rebuild the MCD? (And then we could do something like what Colin suggested below of supporting both the old and new nmstate code in the MCO.) That said we do want to require some of the new nmstate code for 4.13 since [it's part of a larger fix](https://github.com/openshift/machine-config-operator/pull/3684) that also needs to land to handle the NIC pinning issue in all cases. So we could also get this in and backported and that'd also served as prep for #3684 (I'm curious though if Colin was alluding to some other kind of breakage in https://github.com/openshift/machine-config-operator/pull/3685#issuecomment-1520545314).

**Comment by sergiordlr** [comment]
verified using IPI on GCP and UPI on OSP

- UPI on OSP, template `functionality-testing/aos-4_14/upi-on-osp/versioned-installer-https_proxy-fips-ovn-etcd_encryption-ci`

OCP was installed OK

Master nodes skip the nmstatectl execution
```
sh-5.1# journalctl -u machine-config-daemon-firstboot.service

Apr 25 06:41:30 sregidor-fix6-knkw4-master-0 systemd[1]: Starting Machine Config Daemon Firstboot...
Apr 25 06:41:30 sregidor-fix6-knkw4-master-0 podman[2589]: time="2023-04-25T06:41:30Z" level=warning msg="Error loading CNI config file /etc/cni/net.d/200-loopback.conflist: error parsin>pr 25 06:41:30 sregidor-fix6-knkw4-master-0 podman[2589]: 
Apr 25 06:41:30 sregidor-fix6-knkw4-master-0 podman[2589]: 2023-04-25 06:41:30.788956115 +0000 UTC m=+0.128289599 container create b2fdf019cc485978cd3770938f84e5fefe1f151b2f285c1aafc6a8d
```

yum based RHEL8 nodes would be added without problems they do not execute firstboot service at all:
```
sh-4.4# journalctl -u machine-config-daemon-firstboot.service
-- Logs begin at Tue 2023-04-25 04:33:28 EDT, end at Tue 2023-04-25 05:28:18 EDT. --
-- No entries --
```

```
$ oc get nodes -l node.openshift.io/os_id=rhcos
NAME                           STATUS   ROLES                  AGE    VERSION
sregidor-fix6-knkw4-master-0   Ready    control-plane,master   174m   v1.26.3+b404935
sregidor-fix6-knkw4-master-1   Ready    control-plane,master   170m   v1.26.3+b404935
sregidor-fix6-knkw4-master-2   Ready    control-plane,master   165m   v1.26.3+b404935
sregidor-fix6-knkw4-worker-0   Ready    worker                 146m   v1.26.3+b404935
sregidor-fix6-knkw4-worker-1   Ready    worker                 146m   v1.26.3+b404935
$ oc get nodes -l node.openshift.io/os_id=rhel
NAME                          STATUS   ROLES    AGE   VERSION
sregidor-fix6-knkw4-rhel-0    Ready    worker   62m   v1.26.3+b404935
sregidor-fix6-knkw4-rhel-1    Ready    worker   62m   v1.26.3+b404935
sregidor-fix6-knkw4-rhel2-0   Ready    worker   20m   v1.26.3+b404935

```


- IPI on GCP template: functionality-testing/aos-4_14/ipi-on-gcp/versioned-installer-customer_vpc

OCP was installed OK

Master nodes skip the nmstatectl execution

In order to force the nmstatectl execution we duplicated a machineconfig set and used a boot image from OCP 4.12 https://github.com/openshift/installer/blob/release-4.12/data/data/coreos/rhcos.json. The execution of nmstatectl was ok, it didn't use the `--inspect` flag:

```
sh-5.1# journalctl -u machine-config-daemon-firstboot.service

Apr 25 09:09:41 sregidor-fix11-rks5l-worker-hck-nzn52.c.openshift-qe.internal systemd[1]: Starting Machine Config Daemon Firstboot...
Apr 25 09:09:41 sregidor-fix11-rks5l-worker-hck-nzn52.c.openshift-qe.internal sh[2136]: sed: can't read /etc/yum.repos.d/*.repo: No such file or directory
Apr 25 09:09:41 sregidor-fix11-rks5l-worker-hck-nzn52.c.openshift-qe.internal podman[2138]: I0425 09:09:41.658519       1 daemon.go:1576] Persisting NIC names for RHEL8 host system
Apr 25 09:09:41 sregidor-fix11-rks5l-worker-hck-nzn52.c.openshift-qe.internal podman[2138]: I0425 09:09:41.658545       1 daemon.go:1581] Running: /usr/bin/nmstatectl persist-nic-names ->
Apr 25 09:09:41 sregidor-fix11-rks5l-worker-hck-nzn52.c.openshift-qe.internal podman[2138]: [2023-04-25T09:09:41Z INFO  nmstatectl::persist_nic] Would persist the interface ens4 with MAC>
Apr 25 09:09:41 sregidor-fix11-rks5l-worker-hck-nzn52.c.openshift-qe.internal podman[2138]: [2023-04-25T09:09:41Z INFO  nmstatectl::persist_nic] Systemd network link file created at /roo>
Apr 25 09:09:41 sregidor-fix11-rks5l-worker-hck-nzn52.c.openshift-qe.internal machine-config-daemon[2203]: I0425 09:09:41.982190    2203 update.go:2011] Running: systemctl daemon-reload
```

/label qe-approved

**Comment by cgwalters** [comment]
/hold
since the nmstate changes are now reverted and this PR would then re-break things

**Comment by jlebon** [comment]
> /hold since the nmstate changes are now reverted and this PR would then re-break things

I think we still want this PR though since it provides a way to ratchet into the new nmstate. The code here is compatible with both the old and new nmstate code.

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -1574,18 +1574,14 @@ func MaybePersistNetworkInterfaces(osRoot string) error {
 
 	if hostos.IsEL8() {
 		glog.Info("Persisting NIC names for RHEL8 host system")
-	} else {
-		// If we're not on RHEL, just output the current state.  We don't strictly need to do
-		// this but it will greatly help debugging and validation.
-		cmd.Args = append(cmd.Args, "--inspect")
```
Comment (cgwalters) [inline pkg/daemon/daemon.go:1580]: Wait, but this will make us persist the NIC names even when we're starting from rhel9, right?

Comment (jlebon) [reply pkg/daemon/daemon.go:1580]: I intended the cmd logic below so that it's under the el8 check now. :)

**Comment by cgwalters** [comment]
/hold cancel
OK, though I think a downside here is that any output logs from NIC persistence are going to be potentially way back in the systemd journal (if not rotated out) and won't be captured by must-gather by default.

But I think that's OK.

**Comment by cgwalters** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-gcp-ovn-rt-upgrade

**Comment by cgwalters** [comment]
@sdodson The goal is really to test upgrades from RHEL8 to RHEL9 here, which should trigger the NIC name logic

**Comment by cgwalters** [comment]
So because we lost `--inspect`, and the output of the command goes to the pod logs (of which only current is captured) we can't easily verify that the logic ran from the CI job logs.  But it didn't fail for reasons related to this PR as far as I can tell.

**Comment by cgwalters** [comment]
We could I guess change things here to echo the logs to the journal too..

**Comment by jlebon** [comment]
Added more logging here! Hoisted the `getIfnamesFromLinkFiles()` function from https://github.com/openshift/machine-config-operator/pull/3684 so we actually print out the interface names.

**Comment by jlebon** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.13-upgrade-from-stable-4.12-e2e-gcp-ovn-rt-upgrade

**Comment by jlebon** [comment]
OK cool, in the MCO logs now we see e.g.:

```
I0501 17:48:11.519628    4233 update.go:2026] Have persisted ifnames for ens4
```

**Comment by jlebon** [comment]
So I think the steps here are something like:
1. Get this PR into master.
2. Update nmstate in 4.14 to the latest.
3. Verify that 4.14 CI is still happy.
4. Backport this PR to 4.13.
5. Update nmstate in 4.13 to the latest.
6. Verify that 4.13 CI is still happy.

One way to pre-verify step 6 before step 5 in CI is to do something similar to what I crudely did in https://github.com/openshift/machine-config-operator/pull/3684 to fetch nmstate from some other location when building the MCD image.

**Comment by sdodson** [comment]
@jlebon if the hope is to land this before 4.13 GA we need to be done with that all tomorrow.

**Comment by jlebon** [comment]
The end goal is really https://github.com/openshift/machine-config-operator/pull/3684. That PR has non-trivial logic and I'd agree we don't want to rush this in this late in the game. I'm fine trying to get it into a z-stream release instead.

Ideally, we'd add a Known issue entry to the 4.13 release notes for this, but it's hard for users to know if they'll be impacted or not since it depends on their specific network card. Though we could mention at least the models we know have caused issues.

**Comment by jlebon** [comment]
How we do want to proceed with this? Can we get it into 4.14 at least so we can test out https://github.com/openshift/machine-config-operator/pull/3684 too? Or did we want to hold it until we actually see that it's a real issue in the wild? I think when we discussed this last, we didn't have good telemetry in that area to know how likely it'd be to have nodes that hit this.

The failure mode here with the case described in https://github.com/openshift/machine-config-operator/pull/3684 is that the machine will reboot into 9.2 and fail to unlock the rootfs, so it should be a pretty loud failure that should stop the rollout. Users can recover by manually rolling back to the previous deployment.

**Pull request merged**